embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/intake/IntakeServiceImpl.kt

```diff
@@ -23,8 +24,8 @@ class IntakeServiceImpl(
     private val shutdownTimeoutMs: Long = 3000,
 ) : IntakeService {
 
-    private var lastSessionCacheAttempt: Future<*>? = null
-    private var lastCacheSessionRef: StoredTelemetryMetadata? = null
+    private var cachingTasks: MutableMap<SupportedEnvelopeType, Future<*>> = ConcurrentHashMap()
+    private var cacheReferences: MutableMap<SupportedEnvelopeType, StoredTelemetryMetadata> = ConcurrentHashMap()
```

Comment:
nit: make these val

Reply:
Additionally consider make a private class that has a Future property + metadata property. That will avoid the possibility of getting into an inconsistent state where a value is in one collection but not another

Reply:
Yeah sounds good. I had originally started along that path but abandoned it because the generic lookups got harder to read than I wanted. But let me try it again to see if I feel different given you've raised this too.